cli/azd/pkg/infra/provisioning/terraform/terraform_provider.go

```diff
@@ -202,6 +202,17 @@ func (t *TerraformProvider) Deploy(ctx context.Context, deployment *DeploymentPl
 	}, nil
 }
 
+func (t *TerraformProvider) WhatIfDeploy(ctx context.Context, deployment *DeploymentPlan) (*DeployPreviewResult, error) {
```

Comment:
I would have expected that if I ran `azd provision --what-if` when Terraform was my IaC provider this would just print the output from `tf plan`.

I don't think that's what is going to happen today with the changes as is.

I'm not 100% sure what exactly we want to do here - I think we want to just end up showing the tf plan output as is (that matches how we handle showing the raw output from terraform when we do a provision), but it's something we should think about.

I also wonder if on the provider interface we should merge `Plan` and `WhatIfDeploy` somehow.  I think the problem is that unlike terraform, where you always compute a plan before applying it, in ARM land you can do a deploy without a what-if operation, so we'd have to figure out a way to communicate into Plan for the ARM case if you actually wanted to do the what-if part or not.

Reply:
> I would have expected that if I ran azd provision --what-if when Terraform was my IaC provider this would just print the output from tf plan.

That's what it is currently happening due to the provision-manager always calling `Pan()` before `Deploy()` . This makes terraform to only run `Plan()` and show that output. (I ran a test).  Since it returns an empty list of changes, azd does't add more output after it is donde.

I don't think merging Plan() and what-if is a good idea.  I see the `bicep-planning` as the actual `bicep build` operation (transform to arm). 

How does it work in the case of Pulumi?

Reply:
Thanks for the explantation.  I'm still a little nervous about the fact that we have a thing on our interface that is specific to one provider, but that's something we can noodle on in the background and think if there's some better design here.  I think what you have here is good for now, so this feedback is not blocking.

> How does it work in the case of Pulumi?

Pulumi works similar-ish to Terraform.  The typical workflow there is to do a `pulumi up` which will run a preview, ask you to accept it, and then apply it.  There's also a `--skip-preview` flag you can pass, and and an explicit `preview` command that just runs the planning step.

Unlike Terraform, the default workflow doesn't encourage you to produce a reified plan that you then apply later (although they now support something like this).

I suspect that if we added Pulumi support, `Plan()` would call `pulumi preview`, `Deploy()` would call `pulumi --skip-preview`, and then `WhatIfDeploy()` would be a no-op here.

Reply:
I see.
Now I agree with you.  There should be only `plan` and `deploy`.  
That requires updating the `provision` model to only call `plan()` when `--preview` flag.  Then moving the actual `plan()` implementations from the providers to be part of `Deploy()` .
Then making `plan()` to be the implementation I added in this PR.

I think it is worth it to have that in a new PR. 
I am happy to do it (I think it's straight forward).  I just think it will be easier to review and see what's changing in a new PR than making this PR to handle both changes